crates/core/src/node/inner/node_executor.rs

```diff
@@ -178,9 +178,23 @@ impl NodeExecutor {
         }
     }
 
-    async fn set_code(&mut self, address: Address, bytecode: Vec<u8>, reply: oneshot::Sender<()>) {
+    async fn set_code(
+        &mut self,
+        address: Address,
+        mut bytecode: Vec<u8>,
+        reply: oneshot::Sender<()>,
+    ) {
         let code_key = get_code_key(&address);
-        let bytecode_hash = BytecodeHash::for_bytecode(&bytecode).value();
+        let marker = BytecodeMarker::detect(&bytecode);
+        let bytecode_hash = match marker {
+            BytecodeMarker::EraVm => BytecodeHash::for_bytecode(&bytecode),
+            BytecodeMarker::Evm => BytecodeHash::for_raw_evm_bytecode(&bytecode),
+        }
+        .value();
+        if marker == BytecodeMarker::Evm {
+            bytecode = pad_evm_bytecode(&bytecode);
+        }
```

Comment:
Arguably we should detect marker (and even compute hash) only once at the API level but not a huge deal